File: C#/lab2/Task4.java
import java.util.Arrays;

public static float PresentList(float budget, float bagVolume, int people,
                                int Npresents, float[] presentVolume, float[] presentPrice) {
    if (people == 0 || Npresents == 0) return 0.0f;

    // Scale float volumes to integers (assumes at most 2 decimal places)
    final int SCALE = 100;
    int V = (int) (bagVolume * SCALE);

    // dp[k][v] = max total price with (count % people == k) and total volume == v
    // A value of -1.0f means this state is unreachable
    float[][] dp = new float[people][V + 1];
    for (float[] row : dp) Arrays.fill(row, -1.0f);
    dp[0][0] = 0.0f; // Base case: 0 items, 0 volume, 0 cost

    // Process each present (0/1 knapsack - one copy per present)
    for (int i = 0; i < Npresents; i++) {
        int vol = Math.round(presentVolume[i] * SCALE);
        float price = presentPrice[i];

        // Skip presents that don't even fit alone
        if (vol > V) continue;

        // Use a fresh copy so each present is only used once
        float[][] ndp = new float[people][V + 1];
        for (float[] row : ndp) Arrays.fill(row, -1.0f);

        for (int k = 0; k < people; k++) {
            for (int v = 0; v <= V; v++) {
                if (dp[k][v] < 0) continue; // State unreachable, skip

                // Option 1: Don't take present i
                if (ndp[k][v] < dp[k][v])
                    ndp[k][v] = dp[k][v];

                // Option 2: Take present i
                int nv = v + vol;
                if (nv <= V) {
                    int nk = (k + 1) % people;
                    float np = dp[k][v] + price;
                    if (ndp[nk][nv] < np)
                        ndp[nk][nv] = np;
                }
            }
        }
        dp = ndp;
    }

    // Find the maximum affordable spend where count is divisible by people
    float result = 0.0f;
    for (int v = 0; v <= V; v++) {
        // dp[0][v] means count % people == 0, i.e., count is a valid multiple
        if (dp[0][v] >= 0 && dp[0][v] <= budget + 1e-6f) {
            result = Math.max(result, dp[0][v]);
        }
    }
    return result;
}

File: Visual C# .NET/lab2/Task4.cs
using System;

namespace Task4
{
    public class Program
    {
        public static void Execute()
        {
            float budget = 183.23f;
            float bagVolume = 64.11f;
            int people = 7;
            int Npresents = 12;

            float [] presentVolume = new float[Npresents];
            float [] presentPrice = new float[Npresents];
                
            presentVolume[0] = 4.53f;
            presentPrice[0] = 12.23f;
            presentVolume[1] = 9.11f;
            presentPrice[1] = 45.03f;
            presentVolume[2] = 4.53f;
            presentPrice[2] = 12.23f;
            presentVolume[3] = 6.00f;
            presentPrice[3] = 32.93f;
            presentVolume[4] = 1.04f;
            presentPrice[4] = 6.99f;
            presentVolume[5] = 0.87f;
            presentPrice[5] = 0.46f;
            presentVolume[6] = 2.57f;
            presentPrice[6] = 7.34f;
            presentVolume[7] = 19.45f;
            presentPrice[7] = 65.98f;
            presentVolume[8] = 65.59f;
            presentPrice[8] = 152.13f;
            presentVolume[9] = 14.14f;
            presentPrice[9] = 7.23f;
            presentVolume[10] = 16.66f;
            presentPrice[10] = 10.00f;
            presentVolume[11] = 13.53f;
            presentPrice[11] = 25.25f;

            float result = PresentList
            (
                budget, 
                bagVolume, 
                people, 
                Npresents, 
                presentVolume, 
                presentPrice
            );

            Console.WriteLine(result);
        }

        public static float PresentList
        (
            float budget, 
            float bagVolume, 
            int people, 
            int Npresents, 
            float [] presentVolume, 
            float [] presentPrice
        )
        {
            return Solve(
                0,
                0,
                0f,
                0f,

                budget,
                bagVolume,
                people,
                Npresents,
                presentVolume,
                presentPrice
            );
        }

        private static float Solve
        (
            int index,
            int count,
            float usedVolume,
            float usedMoney,

            float budget,
            float bagVolume,
            int people,
            int Npresents,
            float[] presentVolume,
            float[] presentPrice
        )
        {
            // invalid branches
            if (usedVolume > bagVolume || usedMoney > budget)
                return float.NegativeInfinity;

            // base case
            if (index == Npresents)
            {
                // must give equal number of presents to each person
                if (count % people == 0)
                    return usedMoney;

                return float.NegativeInfinity;
            }

            // option 1: skip current present
            float skip = Solve(
                index + 1,
                count,
                usedVolume,
                usedMoney,

                budget,
                bagVolume,
                people,
                Npresents,
                presentVolume,
                presentPrice
            );

            // option 2: take current present
            float take = Solve(
                index + 1,
                count + 1,
                usedVolume + presentVolume[index],
                usedMoney + presentPrice[index],

                budget,
                bagVolume,
                people,
                Npresents,
                presentVolume,
                presentPrice
            );

            return Math.Max(skip, take);
        }
    }
}